### ISSUE TITLE
[Feature Request] `logRetentionDuration`: only delete log files and checkpoint files if not required by a version younger than `logRetentionDuration`

### ISSUE THREAD

**Issue by keen85** [delta-io/delta#1728]
### Overview

As specified in the documentation, there are scenarios where time travel to a previous version is not possible even tough a version is younger than the threshold specified by `delta.logRetentionDuration` and data files are present ([see "_Note_" box in documentation](https://docs.delta.io/latest/delta-batch.html#-data-retention)).
I propose making the logic for selecting and deleting log files and checkpoint files smarter, so that versions younger than `delta.logRetentionDuration` will be readable (as long as underlying data files were not deleted by `VACUUM`).

### Motivation

Current configuration of data retention might indicate to inexperienced users that time traveling to versions younger than the minimum of `delta.logRetentionDuration` and `delta.deletedFileRetentionDuration` will _always_ be possible.
However, there are edge cases where this is not true and where restoring a prior version after accidental operations using time travel will not work.

### Further details

See discussion on Slack: https://delta-users.slack.com/archives/CJ70UCSHM/p1682613236651509

When a new checkpoint is written, deletion of log files and checkpoint files _older_ than `delta.logRetentionDuration` is carried out.
I propose introducing another condition here, checking if these deletion candidates are required by any version that is _younger_ than `delta.logRetentionDuration`.

### Willingness to contribute

The Delta Lake Community encourages new feature contributions. Would you or another member of your organization be willing to contribute an implementation of this feature?

- [ ] Yes. I can contribute this feature independently.
- [ ] Yes. I would be willing to contribute this feature with guidance from the Delta Lake community.
- [x] No. I cannot contribute this feature at this time.

**Comment by scottsand-db** [comment]
@prakharjain09 - do you think you could help here?

**Comment by prakharjain09** [comment]
Yes this seems like a good idea. Instead of deleting every entry prior to the 30 day mark (`delta.logRetentionDuration`), We could identify the newest checkpoint before the 30 day boundary and delete everything before that.

**Comment by keen85** [comment]
Is calculating and writing a checkpoint _expensive_?
If not, why not calc and create an additional checkpoint for the _oldest_ version that is kept (the oldest version that is younger than `delta.logRetentionDuration`), if it does not exists already?

**Comment by ryan-johnson-databricks** [comment]
It actually _can_ be quite expensive for large tables (I've seen 2+ minutes), but it also opens up the possibility of racing checkpoint writers, if two clients are both trying to cleanup at the same time, and both try to write that checkpoint at the same time. See https://github.com/delta-io/delta/issues/1727#issuecomment-1542870175 for an ongoing discussion about what can go wrong when there are races...

**Comment by felipepessoto** [comment]
Duplicate of #606?

**Comment by felipepessoto** [comment]
Done by https://github.com/delta-io/delta/pull/4146
